File: sinais/scripts/analyze_ma12x30_trades_study.mjs
/**
 * Estudo MA12×MA30 offline: Σ simples (A/B), equity composta, max DD, Kelly aprox. (Secção C).
 * `python analyze_ma12x30_trades_study.py` mantém só A/B (sem datas/cronologia no .py por defeito).
 */
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));

function parseEuNum(s) {
  let t = s.trim().replace("$", "").replace(/\s/g, "");
  if (!t) return NaN;
  if (t.includes(",") && t.includes("."))
    return parseFloat(t.replace(/\./g, "").replace(",", "."));
  if (t.includes(",")) return parseFloat(t.replace(",", "."));
  return parseFloat(t);
}

function pctFromLine(line) {
  const m = line.match(/(-?\s*\d+(?:[\.,]\d+)?)%/);
  if (!m) return null;
  return parseFloat(m[1].replace(/\s/g, "").replace(",", "."));
}

function readPriceToken(ln) {
  const tokens = ln.trim().split(/[\t\s]+/);
  for (const tok of tokens) {
    const t = tok.replace("$", "").trim();
    if (/^-?\d/.test(t))
      try {
        return parseEuNum(t);
      } catch (_) {}
  }
  return parseEuNum(ln.replace("$", ""));
}

/** `dd/mm/yyyy` ou `dd/mm/yyyy hh:mm` → ms UTC; `null` se não casar. */
function parsePtDateTime(line) {
  const m = line
    ?.trim()
    .match(/^(\d{2})\/(\d{2})\/(\d{4})(?:\s+(\d{1,2}):(\d{2}))?$/);
  if (!m) return null;
  const d = Number(m[1]);
  const mo = Number(m[2]);
  const y = Number(m[3]);
  const h = m[4] != null ? Number(m[4]) : 12;
  const mi = m[5] != null ? Number(m[5]) : 0;
  const t = Date.UTC(y, mo - 1, d, h, mi, 0, 0);
  return Number.isFinite(t) ? t : null;
}

function parseTradesBlob(text) {
  const lines = text.split(/\n/).map((l) => l.trim()).filter(Boolean);
  /** @typedef {{symbol:string,side:string,entry:number,exit_px:number,mfe_pct:number,mae_pct:number,actual_ret_pct:number,ts:number|null,fileOrder:number}} Trade */
  const trades = /** @type {Trade[]} */ ([]);
  let fileOrder = 0;
  const symRe = /^([A-Z0-9]+USDT)\s+(COMPRA|VENDA)/;
  let i = 0;
  while (i < lines.length) {
    const m = lines[i].match(symRe);
    if (!m) {
      i++;
      continue;
    }
    const symbol = m[1];
    const side = m[2];
    let j = i + 1;
    while (
      j < lines.length &&
      !lines[j].startsWith("15m\t") &&
      !lines[j].startsWith("15m ")
    )
      j++;
    if (j >= lines.length) break;
    const row = lines[j];
    const partsTab = row.split("\t").filter(Boolean);
    const ents = [];
    for (const chunk of partsTab.slice(1)) {
      const cc = chunk.replace("$", "").trim();
      if (!cc) continue;
      try {
        ents.push(parseEuNum(cc));
      } catch (_) {}
    }
    let entry, exit_px;
    if (ents.length >= 2) {
      entry = ents[ents.length - 2];
      exit_px = ents[ents.length - 1];
    } else {
      const dp = [...row.matchAll(/(\d+[.,]\d+)/g)];
      if (dp.length < 2) {
        i++;
        continue;
      }
      entry = parseEuNum(dp[dp.length - 2][1]);
      exit_px = parseEuNum(dp[dp.length - 1][1]);
    }

    let k = j + 1;
    // Linha monetária do PnL com $ (ex.: $-8,22 ou $+15,71)
    if ((lines[k] ?? "").includes("$")) k++;
    let act =
      side === "COMPRA"
        ? ((exit_px - entry) / entry) * 100
        : ((entry - exit_px) / entry) * 100;
    const pct = pctFromLine(lines[k] ?? "");
    if (pct != null) {
      act = pct;
      k++;
    }

    const hi_px = readPriceToken(lines[k] ?? "");
    k++;
    if (pctFromLine(lines[k] ?? "") != null) k++;
    const lo_px = readPriceToken(lines[k] ?? "");
    k++;
    if (pctFromLine(lines[k] ?? "") != null) k++;

    let mfe_pct, mae_pct;
    if (side === "COMPRA") {
      mfe_pct = ((hi_px - entry) / entry) * 100;
      mae_pct = ((lo_px - entry) / entry) * 100;
    } else {
      mfe_pct = ((entry - lo_px) / entry) * 100;
      mae_pct = ((entry - hi_px) / entry) * 100;
    }

    let ts = null;
    const parsedTs = parsePtDateTime(lines[k] ?? "");
    if (parsedTs != null) {
      ts = parsedTs;
      k++;
    }

    trades.push({
      symbol,
      side,
      entry,
      exit_px,
      high: hi_px,
      low: lo_px,
      mfe_pct,
      mae_pct,
      actual_ret_pct: act,
      ts,
      fileOrder: fileOrder++,
    });

    while (k < lines.length && !symRe.test(lines[k])) k++;
    i = k;
  }
  return trades;
}

function pnlWithPartials(t, sl_pct, tp1, tp2, w1, w2) {
  const hit_sl = t.mae_pct <= -sl_pct;
  if (hit_sl) return -sl_pct;
  let rest = 1.0;
  let total = 0.0;
  if (t.mfe_pct >= tp1 && rest > 1e-9) {
    const take = Math.min(w1, rest);
    total += take * tp1;
    rest -= take;
  }
  if (t.mfe_pct >= tp2 && rest > 1e-9) {
    const take = Math.min(w2, rest);
    total += take * tp2;
    rest -= take;
  }
  total += rest * t.actual_ret_pct;
  return total;
}

/** Ordem temporal ascendente; sem `ts` mantém ordem no ficheiro (`fileOrder`). */
function sortChronological(arr) {
  return [...arr].sort((a, b) => {
    if (a.ts != null && b.ts != null) return a.ts - b.ts;
    if (a.ts != null) return -1;
    if (b.ts != null) return 1;
    return a.fileOrder - b.fileOrder;
  });
}

/**
 * @param {number[]} pctSeries - retorno % por trade (sequência cronológica)
 */
function seriesStatsFromReturns(pctSeries) {
  if (!pctSeries.length) {
    return {
      finalEq: 100,
      maxDdPct: 0,
      sumSimple: 0,
      kellyMerton: 0,
      kellyBinary: 0,
      muPct: 0,
      sigmaPct: 0,
      n: 0,
    };
  }
  let eq = 100;
  let peak = 100;
  let maxDdPct = 0;
  for (let i = 0; i < pctSeries.length; i++) {
    const r = pctSeries[i];
    eq *= 1 + r / 100;
    if (eq > peak) peak = eq;
    const dd = peak > 0 ? ((peak - eq) / peak) * 100 : 0;
    if (dd > maxDdPct) maxDdPct = dd;
  }
  const rs = pctSeries.map((x) => x / 100);
  const n = rs.length;
  const mu = rs.reduce((s, x) => s + x, 0) / n;
  const var_ = rs.reduce((s, x) => s + (x - mu) ** 2, 0) / n;
  const sigma = Math.sqrt(Math.max(var_, 0));
  const kellyMerton = var_ > 1e-16 ? mu / var_ : 0;
  const pos = pctSeries.filter((x) => x > 0);
  const neg = pctSeries.filter((x) => x < 0);
  const p = pos.length / n;
  const avgWin = pos.length ? pos.reduce((a, b) => a + b, 0) / pos.length : 0;
  const avgLossAbs = neg.length ? neg.reduce((a, b) => a + Math.abs(b), 0) / neg.length : 0;
  let kellyBinary = 0;
  if (avgWin > 0 && avgLossAbs > 0) {
    const b = avgWin / avgLossAbs;
    kellyBinary = p - (1 - p) / b;
  }
  return {
    finalEq: eq,
    maxDdPct,
    sumSimple: pctSeries.reduce((s, x) => s + x, 0),
    kellyMerton,
    kellyBinary,
    muPct: mu * 100,
    sigmaPct: sigma * 100,
    n,
  };
}

function summarize(trades) {
  let actSum = trades.reduce((s, t) => s + t.actual_ret_pct, 0);
  const sumMfe = trades.reduce((s, t) => s + t.mfe_pct, 0);
  const wins = trades.filter((t) => t.actual_ret_pct > 0).length;
  console.log(
    `Trades: ${trades.length} | soma retornos observados: ${actSum.toFixed(2)}% | média ${(actSum / trades.length).toFixed(3)}%`
  );
  console.log(
    `Win rate observado: ${wins}/${trades.length} (${((100 * wins) / trades.length).toFixed(1)}%)`
  );
  const mfeNeg = trades.filter((t) => t.mfe_pct < 0).length;
  console.log(
    `Soma MFE (%): ${sumMfe.toFixed(0)} | trades com MFE<0: ${mfeNeg} (LONG só em queda líquida do máximo)`
  );

  const slGrid = [3, 4, 5, 6, 7, 8, 10, 12, 15];
  const tpPairs = [
    [6, 12],
    [8, 16],
    [10, 20],
    [12, 25],
    [15, 30],
    [20, 40],
    [30, 50],
    [44, 44],
  ];
  const weights = [
    [0.4, 0.4],
    [0.5, 0.3],
    [0.6, 0.3],
  ];

  console.log(
    "\n--- A: restante ao fecho real (proxy estratégia / time stop até ao fecho listado) ---\n"
  );
  for (const [w1, w2] of weights) {
    for (const [tp1, tp2] of tpPairs) {
      if (tp2 < tp1) continue;
      const parts = [];
      for (const sl of slGrid) {
        const s = trades.reduce(
          (acc, t) => acc + pnlWithPartials(t, sl, tp1, tp2, w1, w2),
          0
        );
        parts.push(`SL${sl}%=${s.toFixed(0)}`);
      }
      console.log(`TP ${tp1}/${tp2} w ${w1}/${w2} | ${parts.join(" | ")}`);
    }
  }

  console.log(
    "\n--- B: remanescente com `max(MFE observado, fecho real)` (teto pessimista/coerente vs duplicar TPs+MFE inteiro) ---\n"
  );
  /** @param {typeof trades[number]} t */
  function pnlMfeRemainderCapped(t, sl_pct, tp1, tp2, w1, w2) {
    if (t.mae_pct <= -sl_pct) return -sl_pct;
    let rest = 1.0;
    let total = 0.0;
    if (t.mfe_pct >= tp1 && rest > 1e-9) {
      const take = Math.min(w1, rest);
      total += take * tp1;
      rest -= take;
    }
    if (t.mfe_pct >= tp2 && rest > 1e-9) {
      const take = Math.min(w2, rest);
      total += take * tp2;
      rest -= take;
    }
    const tail = Math.max(t.mfe_pct, t.actual_ret_pct);
    total += rest * tail;
    return total;
  }
  for (const [w1, w2] of weights.slice(0, 2)) {
    for (const [tp1, tp2] of [[15, 30], [20, 40], [44, 44]]) {
      const row = slGrid.map((sl) => {
        const s = trades.reduce(
          (acc, t) =>
            acc + pnlMfeRemainderCapped(t, sl, tp1, tp2, w1, w2),
          0
        );
        return `${s.toFixed(0)}`;
      });
      console.log(`TP ${tp1}/${tp2} w ${w1}/${w2}: SL grid → ${row.join(", ")}`);
    }
  }

  console.log("\n--- Topo por soma total (cenário A, todos SL da grelha) ---");
  const allConfigs = [];
  for (const [w1, w2] of weights)
    for (const [tp1, tp2] of tpPairs)
      for (const sl of slGrid) {
        const s = trades.reduce(
          (acc, t) => acc + pnlWithPartials(t, sl, tp1, tp2, w1, w2),
          0
        );
        allConfigs.push({ s, sl, tp1, tp2, w1, w2 });
      }
  allConfigs.sort((a, b) => b.s - a.s);
  const seen = new Set();
  for (const c of allConfigs) {
    const k = `${c.sl}-${c.tp1}-${c.tp2}-${c.w1}-${c.w2}`;
    if (seen.has(k)) continue;
    seen.add(k);
    console.log(
      `Σ=${c.s.toFixed(0)}%  SL=${c.sl}%  TP= ${c.tp1}/${c.tp2}%  w= ${c.w1}/${c.w2}`
    );
    if (seen.size >= 15) break;
  }

  console.log("\n--- Config produto aprox.: 1 TP 44% em 60% posição ---");
  for (const sl of slGrid) {
    const s = trades.reduce(
      (acc, t) => acc + pnlWithPartials(t, sl, 44, 44, 0.6, 0),
      0
    );
    console.log(`  SL ${sl}%  → Σ ${s.toFixed(0)}%`);
  }

  const slTriggers = {};
  for (const sl of [5, 7, 10]) {
    slTriggers[sl] = trades.filter((t) => t.mae_pct <= -sl).length;
  }

  const dated = trades.filter((t) => t.ts != null).length;
  console.log(
    `\n--- C: Equity composta (base 100), max drawdown, Kelly aproximado ---\n` +
      `Datas parseadas: ${dated}/${trades.length} (ordem temporal ascendente).\n` +
      `Kelly Merton ≈ μ/σ² (retornos % como fraccionários); Kelly binário = p − (1−p)/b, b=avgWin/avgLoss. ` +
      `Valores f* >> 100% indicam modelo i.i.d. frágil (ignorar como tamanho de posição literal). ` +
      `Preferir fração << f* (ex. meio-Kelly); amostra curta e não estacionária.\n`
  );

  function fmtRisk(st) {
    return (
      `Eq→${st.finalEq.toFixed(2)} | maxDD ${st.maxDdPct.toFixed(2)}% | ` +
      `μ ${st.muPct.toFixed(3)}% σ ${st.sigmaPct.toFixed(3)}% | ` +
      `f*_M ${st.kellyMerton.toFixed(4)} (½ ${(st.kellyMerton / 2).toFixed(4)}) | ` +
      `f*_B ${st.kellyBinary.toFixed(4)} (½ ${(st.kellyBinary / 2).toFixed(4)})`
    );
  }

  const stObs = seriesStatsFromReturns(trades.map((t) => t.actual_ret_pct));
  console.log("[Observado — fechos reais] " + fmtRisk(stObs));

  console.log("\n[Produto sintét.: TP +44% em 60%]");
  const prodRows = [];
  for (const sl of slGrid) {
    const srs = trades.map((t) => pnlWithPartials(t, sl, 44, 44, 0.6, 0));
    const st = seriesStatsFromReturns(srs);
    prodRows.push({ sl, st });
    console.log(`  SL ${String(sl).padStart(2)}%  ${fmtRisk(st)}`);
  }

  console.log("\n[Simul.: TP 44/44, 40%+40%, resto fecho real]");
  for (const sl of slGrid) {
    const srs = trades.map((t) => pnlWithPartials(t, sl, 44, 44, 0.4, 0.4));
    const st = seriesStatsFromReturns(srs);
    console.log(`  SL ${String(sl).padStart(2)}%  ${fmtRisk(st)}`);
  }

  console.log("\n--- SL vs equity final vs max DD (produto sintét. 44%×60%) ---");
  console.log(prodRows.map((r) => `SL${r.sl}% Eq${r.st.finalEq.toFixed(1)} DD${r.st.maxDdPct.toFixed(1)}%`).join(" | "));
  const bestEq = [...prodRows].sort((a, b) => b.st.finalEq - a.st.finalEq)[0];
  console.log(`Maior equity final nesta amostra: SL=${bestEq.sl}% (${bestEq.st.finalEq.toFixed(2)})`);

  console.log(
    `\n--- Trades com MAE que tocou SL (referência) --- ${JSON.stringify(slTriggers)}`
  );
  console.log(
    "MAE só até ao fecho listado; Kelly é ilustrativo, não aconselhamento de posição."
  );
}

const blobPath =
  process.argv[2] || path.join(__dirname, "ma12x30_trades_paste.txt");
const blob = fs.readFileSync(blobPath, "utf8");
const trades = sortChronological(parseTradesBlob(blob));
if (trades.length < 10) {
  console.error("Poucos trades:", trades.length);
  process.exit(1);
}
summarize(trades);

File: sinais/scripts/analyze_ma12x30_trades_study.py
"""
Estudo offline: trades MA12×MA30 (dados colados da UI).
Usa entrada, saída real, máxima e mínima intradiárias (MFE/MAE em %).

Modelo de encadeamento (conservador para SL):
- Se o MAE tocou o SL (% vs entrada, no sentido da posição), assume-se saída total no SL.
- Caso contrário: aplica TP1 em fração w1 se MFE >= TP1%; depois TP2 em fração w2 do resto se MFE >= TP2%;
  o remanescente fecha na % do preço real de saída (proxy de fecho dinâmico / 24h).

Variante "24h": igual, mas o remanescente usa sempre o retorno do fecho observado (já é o caso);
  para comparar com "sem time stop" podíamos usar só MFE no remanescente — aqui "24h" = assumir
  que o fecho listado é a marcação após janela (mesmo número que baseline real).

Limitação: sem série temporal intra-trade, não sabemos se o SL foi tocado antes ou depois dos TPs;
  a regra acima dá prioridade implícita ao SL se MAE <= -SL (realista para stops fixos).

Equity composta, max drawdown e Kelly (ordem cronológica com datas dd/mm/yyyy): ver
`sinais/scripts/analyze_ma12x30_trades_study.mjs` (Node).
"""
from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Literal

Side = Literal["COMPRA", "VENDA"]


@dataclass
class Trade:
    symbol: str
    side: Side
    entry: float
    exit_px: float
    high: float
    low: float
    mfe_pct: float  # melhor movimento a favor (% PnL vs entrada)
    mae_pct: float  # pior movimento contra (% PnL vs entrada, negativo em long)
    actual_ret_pct: float


def parse_eu_num(s: str) -> float:
    s = s.strip().replace("$", "").replace(" ", "")
    if "," in s and "." in s:
        # e.g. 1.234,56
        s = s.replace(".", "").replace(",", ".")
    elif "," in s:
        s = s.replace(",", ".")
    return float(s)


def _pct_from_line(line: str) -> float | None:
    m = re.search(r"(-?\s*\d+(?:[\.,]\d+)?)%", line)
    if not m:
        return None
    return float(m.group(1).replace(" ", "").replace(",", "."))


def _read_price_token(ln: str) -> float:
    """Primeiro token tipo preço (pode vir '0,0007103' ou '13,054')."""
    ln = ln.strip()
    tokens = re.split(r"[\t\s]+", ln)
    for tok in tokens:
        t = tok.strip().replace("$", "")
        if t and re.match(r"^-?\d", t):
            try:
                return parse_eu_num(t)
            except ValueError:
                continue
    return parse_eu_num(ln.replace("$", ""))


def parse_trades_blob(text: str) -> list[Trade]:
    """Parser ao bloco colado: SYMBOL + lado → linha 15m → PnL% → dois preços+cada %."""
    lines = [ln.strip() for ln in text.splitlines() if ln.strip()]
    trades: list[Trade] = []
    sym_re = re.compile(r"^([A-Z0-9]+USDT)\s+(COMPRA|VENDA)")

    i = 0
    while i < len(lines):
        m = sym_re.match(lines[i])
        if not m:
            i += 1
            continue
        symbol = m.group(1)
        side = m.group(2)  # type: Side
        j = i + 1
        while j < len(lines) and not (lines[j].startswith("15m\t") or lines[j].startswith("15m ")):
            j += 1
        if j >= len(lines):
            break
        row = lines[j]
        parts_tab = row.split("\t")
        ents: list[float] = []
        for chunk in parts_tab[1:]:
            cc = chunk.replace("$", "").strip()
            if not cc:
                continue
            try:
                ents.append(parse_eu_num(cc))
            except ValueError:
                pass
        if len(ents) >= 2:
            entry = ents[-2]
            exit_px = ents[-1]
        else:
            dp = list(re.finditer(r"(\d+[.,]\d+)", row))
            if len(dp) < 2:
                i += 1
                continue
            entry = parse_eu_num(dp[-2].group(1))
            exit_px = parse_eu_num(dp[-1].group(1))

        k = j + 1
        # Linha monetária do PnL com símbolo $ (mesmo quando negativa, ex.: $-8,22)
        if k < len(lines) and "$" in lines[k]:
            k += 1
        # Retorno observado %
        act = (
            ((exit_px - entry) / entry) * 100
            if side == "COMPRA"
            else ((entry - exit_px) / entry) * 100
        )
        if k < len(lines) and _pct_from_line(lines[k]) is not None:
            act = float(_pct_from_line(lines[k]) or act)
            k += 1

        if k >= len(lines):
            break
        hi_px = _read_price_token(lines[k])
        k += 1
        if k < len(lines) and _pct_from_line(lines[k]) is not None:
            k += 1
        if k >= len(lines):
            break
        lo_px = _read_price_token(lines[k])
        k += 1
        if k < len(lines) and _pct_from_line(lines[k]) is not None:
            k += 1

        # MFE / MAE = retorno % da posição no melhor/pior extreme (vs entrada)
        if side == "COMPRA":
            mfe_pct = ((hi_px - entry) / entry) * 100
            mae_pct = ((lo_px - entry) / entry) * 100
        else:
            mfe_pct = ((entry - lo_px) / entry) * 100
            mae_pct = ((entry - hi_px) / entry) * 100

        trades.append(
            Trade(
                symbol=symbol,
                side=side,
                entry=entry,
                exit_px=exit_px,
                high=hi_px,
                low=lo_px,
                mfe_pct=mfe_pct,
                mae_pct=mae_pct,
                actual_ret_pct=act,
            )
        )

        while k < len(lines) and not sym_re.match(lines[k]):
            k += 1
        i = k

    return trades


def pnl_with_partials(
    t: Trade,
    sl_pct: float,
    tp1: float,
    tp2: float,
    w1: float,
    w2: float,
) -> float:
    """Retorno % sobre a posição total ( média ponderada das fatias )."""
    if t.side == "COMPRA":
        hit_sl = t.mae_pct <= -sl_pct
    else:
        hit_sl = t.mae_pct <= -sl_pct

    if hit_sl:
        return -sl_pct

    rest = 1.0
    total = 0.0
    # TP1
    if t.mfe_pct >= tp1 and rest > 1e-9:
        take = min(w1, rest)
        total += take * tp1
        rest -= take
    # TP2 (fracção do original position)
    if t.mfe_pct >= tp2 and rest > 1e-9:
        take = min(w2, rest)
        total += take * tp2
        rest -= take
    # resto ao fecho real
    total += rest * t.actual_ret_pct
    return total


def summarize(trades: list[Trade]) -> None:
    act = sum(t.actual_ret_pct for t in trades)
    print(f"Trades: {len(trades)} | soma retornos observados %: {act:.2f}% | média {act/len(trades):.3f}%")
    wins = sum(1 for t in trades if t.actual_ret_pct > 0)
    print(f"Win rate observado: {wins}/{len(trades)} = {100*wins/len(trades):.1f}%")

    sl_grid = [3, 4, 5, 6, 7, 8, 10, 12, 15]
    # Dois TP parciais — grelhas (TP1 menor, TP2 maior); pesos típicos 40/40/20 runners
    tp_pairs = [(6, 12), (8, 16), (10, 20), (12, 25), (15, 30), (20, 40), (30, 50), (44, 44)]
    weights = [(0.40, 0.40), (0.50, 0.30), (0.60, 0.30)]

    print("\n--- Grelha: SL × (TP1,TP2,w1,w2), restante no fecho real (proxy 24h/dinâmico) ---\n")

    best: list[tuple[float, tuple]] = []

    for w1, w2 in weights:
        for tp1, tp2 in tp_pairs:
            if tp2 < tp1:
                continue
            row = []
            for sl in sl_grid:
                s = sum(pnl_with_partials(t, sl, tp1, tp2, w1, w2) for t in trades)
                row.append((sl, s))
                best.append((s, (sl, tp1, tp2, w1, w2)))
            line = ", ".join(f"SL{s:g}%→Σ{smm:.0f}" for s, smm in row)
            print(f"TP1={tp1}% TP2={tp2}%  w=( {w1:.0f} , {w2:.0f} ) | {line}")

    best.sort(reverse=True, key=lambda x: x[0])
    print("\n--- Top 12 combinações por soma total % ---")
    seen = set()
    for total, cfg in best:
        key = cfg
        if key in seen:
            continue
        seen.add(key)
        print(f"Σ={total:.1f}%  SL={cfg[0]}%  TP1={cfg[1]}% TP2={cfg[2]}%  w1={cfg[3]:.2f} w2={cfg[4]:.2f}")
        if len(seen) >= 12:
            break

    # Produto actual aproximado: um TP a 44% 60% posição (só 1 parcial no código)
    print("\n--- Aproximação config produto: 1 TP a 44% em 60% posição ---")
    for sl in sl_grid:
        s_single = sum(
            pnl_with_partials(t, sl, 44.0, 44.0, 0.60, 0.0) for t in trades
        )
        print(f"  SL {sl}%  → soma Σ {s_single:.1f}%")


# Colar dados abaixo (entre TRADES_BEGIN / END) quando correr standalone.
TRADES_PLACEHOLDER = "TRADES_BEGIN\nTRADES_END"


if __name__ == "__main__":
    import sys
    from pathlib import Path

    path = Path(__file__).resolve()
    blob_file = path.parent / "ma12x30_trades_paste.txt"
    blob = blob_file.read_text(encoding="utf-8") if blob_file.exists() else ""
    if not blob.strip():
        blob = sys.stdin.read()
    trades = parse_trades_blob(blob)
    if len(trades) < 50:
        print("Poucos trades parseados;", len(trades), "— verificar ficheiro ma12x30_trades_paste.txt")
        if trades:
            for x in trades[:3]:
                print(x)
        sys.exit(1)
    summarize(trades)
